### PR TITLE
added order service and order controller

### PR DESCRIPTION
**Pull request by kriazantseva**

### CODE REVIEW

File: src/main/java/mate/academy/bookstoreapp/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    public OrderDto getOrder(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return orderService.getOrdersByUserId(user.getId());
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Change order status", description = "Change order status")
+    @ResponseStatus(HttpStatus.OK)
+    @PatchMapping("/{orderId}")
+    public void updateOrderStatus(@PathVariable Long orderId,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/controller/OrderController.java]:
```suggestion
    public OrderDto updateOrderStatus(@PathVariable Long orderId,
```

File: src/main/java/mate/academy/bookstoreapp/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -72,4 +73,15 @@ public void cleanCart(Authentication authentication) {
         User user = (User) authentication.getPrincipal();
         shoppingCartService.clearCart(user.getId());
     }
+
+    @Operation(summary = "Make an order", description = "Transform cart to order")
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping("/order")
+    public void makeOrder(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/controller/ShoppingCartController.java]: all logic related to order should be in OrderController

File: src/main/java/mate/academy/bookstoreapp/dto/order/CreateOrderRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package mate.academy.bookstoreapp.dto.order;
+
+import jakarta.validation.constraints.NotEmpty;
+
+public record CreateOrderRequestDto(
+        @NotEmpty String shippingAddress
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/dto/order/CreateOrderRequestDto.java]:
```suggestion
        @NotBlank String shippingAddress
```

File: src/main/java/mate/academy/bookstoreapp/dto/orderitem/CreateOrderItemRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.bookstoreapp.dto.orderitem;
+
+import jakarta.validation.constraints.Max;
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record CreateOrderItemRequestDto(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/dto/orderitem/CreateOrderItemRequestDto.java:10]: Where do you use this dto?

File: src/main/java/mate/academy/bookstoreapp/model/CartItem.java [hunk 5]
```diff
@@ -9,13 +9,17 @@
...
 import lombok.Getter;
 import lombok.Setter;
+import lombok.ToString;
 
 @Entity
 @Table(name = "cart_items")
 @Getter
 @Setter
+@ToString
+@EqualsAndHashCode
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/model/CartItem.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstoreapp/model/CartItem.java [hunk 6]
```diff
@@ -9,13 +9,17 @@
...
 public class CartItem {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/model/CartItem.java:22]: add `@ManyToOne(fetch = FetchType.LAZY, optional = false)` for private ShoppingCart shoppingCart;

File: src/main/java/mate/academy/bookstoreapp/model/Order.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+@Table(name = "orders")
+@Getter
+@Setter
+@ToString(exclude = "orderItems")
+@EqualsAndHashCode(exclude = {"orderItems", "status"})
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/model/Order.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/bookstoreapp/model/Order.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Status status;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/model/Order.java]:
```suggestion
    private Status status = Status.PENDING;
```

you can initialize it straightaway

File: src/main/java/mate/academy/bookstoreapp/model/Order.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    @Column(name = "order_date", nullable = false)
+    private LocalDateTime orderDate = LocalDateTime.now();
+    @Column(name = "shipping_address", nullable = false)
+    private String shippingAddress;
+    @OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST, orphanRemoval = true)
+    @JoinTable(
+            name = "orders_order_items",
+            joinColumns = @JoinColumn(name = "order_id"),
+            inverseJoinColumns = @JoinColumn(name = "order_item_id")
+    )
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/model/Order.java]:
```suggestion
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/mate/academy/bookstoreapp/model/Order.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Status status;
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(name = "order_date", nullable = false)
+    private LocalDateTime orderDate = LocalDateTime.now();
+    @Column(name = "shipping_address", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/model/Order.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/mate/academy/bookstoreapp/model/Order.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    private Long id;
+    @ManyToOne
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
+    private Status status;
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(name = "order_date", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/model/Order.java]:
```suggestion
    @Column(nullable = false)

```

File: src/main/java/mate/academy/bookstoreapp/model/OrderItem.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+@Getter
+@Setter
+@ToString
+@EqualsAndHashCode
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "order_id", nullable = false)
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/model/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    public OrderItemDto getOrderItemById(Long userId, Long orderId, Long orderItemId) {
+        Order order = findOrderByUserIdAndOrderId(userId, orderId);
+        Set<OrderItem> orderItems = order.getOrderItems();
+        for (OrderItem orderItem : orderItems) {
+            if (orderItem.getId().equals(orderItemId)) {
+                return orderItemMapper.toDto(orderItem);
+            }
+        }
+        throw new EntityNotFoundException("Item with id: " + orderItemId + " had no order items");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/service/impl/OrderServiceImpl.java]:
```suggestion
        OrderItem orderItem = itemRepository.findByIdAndOrderIdAndUserId(orderItemId, orderId, userId).orElseThrow(() -> new EntityNotFoundException("Item with id: " + orderItemId + " had no order items");
        throw new EntityNotFoundException("Item with id: " + orderItemId + " had no order items");
        return orderItemMapper.toDto(orderItem);
```

File: src/main/java/mate/academy/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+@Transactional
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+
+    @Override
+    public OrderDto getOrdersByUserId(Long userId) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/service/impl/OrderServiceImpl.java]:
```suggestion
    public OrderDto getOrderByUserId(Long userId) {
```

File: src/main/java/mate/academy/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+import org.springframework.stereotype.Service;
+
+@Transactional
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/service/impl/OrderServiceImpl.java:41]: add getAllOrders method

File: src/main/java/mate/academy/bookstoreapp/service/impl/ShoppingCartServiceImpl.java [hunk 16]
```diff
@@ -73,6 +86,25 @@ public void clearCart(Long userId) {
         shoppingCartRepository.save(shoppingCart);
     }
 
+    @Override
+    public void transformCartToOrder(Long id, CreateOrderRequestDto requestDto) {
+        Order order = orderMapper.toEntity(requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/service/impl/ShoppingCartServiceImpl.java]: create method createOrder in OrderServiceImpl and implement it there 
as I understood these logic related to the create order

File: src/main/java/mate/academy/bookstoreapp/service/impl/OrderServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+import mate.academy.bookstoreapp.dto.orderitem.OrderItemDto;
+import mate.academy.bookstoreapp.mapper.OrderItemMapper;
+import mate.academy.bookstoreapp.mapper.OrderMapper;
+import mate.academy.bookstoreapp.model.Order;
+import mate.academy.bookstoreapp.model.OrderItem;
+import mate.academy.bookstoreapp.repository.order.OrderRepository;
+import mate.academy.bookstoreapp.service.OrderService;
+import org.springframework.stereotype.Service;
+
+@Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstoreapp/service/impl/OrderServiceImpl.java]: should we use this annotation for all methods in service?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**